01-architecture.md

```diff
@@ -0,0 +1,222 @@
+# Filecoin VM architecture
+
+This documents provides an overview of the architecture of the FVM, including a small summary of the current VM as an annex at the end.
+
+<!-- START doctoc generated TOC please keep comment here to allow auto update -->
+<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
+**Table of Contents**  *generated with [DocToc](https://github.com/thlorenz/doctoc)*
+
+- [Overview](#overview)
+- [Native VM](#native-vm)
+- [Built-in system actors](#built-in-system-actors)
+- [Actor sandbox](#actor-sandbox)
+- [Actor public interface](#actor-public-interface)
+- [IPLD everything](#ipld-everything)
+- [State access](#state-access)
+- [Chain access](#chain-access)
+- [Actor deployment](#actor-deployment)
+- [Cross-actor calls](#cross-actor-calls)
+- [Call patterns](#call-patterns)
+- [Syscalls](#syscalls)
+- [Gas accounting](#gas-accounting)
+- [WASM interpreters](#wasm-interpreters)
+- [JSON-RPC API](#json-rpc-api)
+- [Interoperability with other networks](#interoperability-with-other-networks)
+- [Formal verifiability](#formal-verifiability)
+- [Upgradability](#upgradability)
+- [Annex: Current VM](#annex-current-vm)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+
+## Overview
+
+![Proposed VM architecture](img/proposed-vm-arch.png)
+
+The FVM aims to (a) support a multitude of programming models for actors and (b) facilitate the onboarding of smart contracts and programs written for other environments, so they can leverage the storage capabilities of the Filecoin network.
+
+Its architecture is inspired by the [hypervisor](https://en.wikipedia.org/wiki/Hypervisor) concept from the world of Virtual Machines. The abstraction is not translated directly, though.
+
+## Native VM
+
+The native runtime will be based on WebAssembly (WASM), with potential usage of a limited set of WebAssembly System Interface (WASI) APIs.
```

Comment:
We had discussed having the SDK interact with the state store directly, which would place the filesystem IO calls in WASM land, necessitating WASI. This is to avoid the FFI barrier cost per every CID lookup.

Reply:
To be clear: WASI doesn't save us any `WASM<->Rust` FFI costs because `WASI` is _implemented_ in rust.

Reply:
@Stebalien and I discussed offline and we came to a common understanding of the boundaries involved, and the events that would lead to boundary traversals. I've captured this in a diagram and in the doc, and will be pushing shortly.